src/content/datachannel/filetransfer/js/main.js

```
@@ -90,91 +90,84 @@ function sendData() {
-  );
+async function gotLocalDescription(desc) {
+  await localConnection.setLocalDescription(desc);
+  console.log(`Offer from localConnection\n ${desc.sdp}`);
+  await remoteConnection.setRemoteDescription(desc);
+  try {
+    const answer = await remoteConnection.createAnswer();
+    await gotRemoteDescription(answer);
+  } catch (e) {
+    onCreateSessionDescriptionError(e);
+  }
 }
 
-function gotDescription2(desc) {
-  remoteConnection.setLocalDescription(desc);
-  trace('Answer from remoteConnection \n' + desc.sdp);
-  localConnection.setRemoteDescription(desc);
+async function gotRemoteDescription(desc) {
+  await remoteConnection.setLocalDescription(desc);
+  console.log(`Answer from remoteConnection\n ${desc.sdp}`);
+  await localConnection.setRemoteDescription(desc);
 }
 
 function getOtherPc(pc) {
   return (pc === localConnection) ? remoteConnection : localConnection;
 }
 
 function getName(pc) {
-  return (pc === localConnection) ? 'localPeerConnection' :
-      'remotePeerConnection';
+  return (pc === localConnection) ? 'localPeerConnection' : 'remotePeerConnection';
 }
 
-function onIceCandidate(pc, event) {
-  getOtherPc(pc).addIceCandidate(event.candidate)
-  .then(
-    function() {
-      onAddIceCandidateSuccess(pc);
-    },
-    function(err) {
-      onAddIceCandidateError(pc, err);
-    }
-  );
-  trace(getName(pc) + ' ICE candidate: \n' + (event.candidate ?
-      event.candidate.candidate : '(null)'));
-}
-
-function onAddIceCandidateSuccess() {
-  trace('AddIceCandidate success.');
-}
-
-function onAddIceCandidateError(error) {
-  trace('Failed to add Ice Candidate: ' + error.toString());
+async function onIceCandidate(pc, event) {
+  try {
+    const candidateText = event.candidate ? event.candidate.candidate : '(null)';
+    console.log(`${getName(pc)} ICE candidate: ${candidateText}`);
+    await getOtherPc(pc).addIceCandidate(event.candidate);
+    console.log('AddIceCandidate success.');
+  } catch (e) {
```

This try/catch will eat the error, so higher level functions will not see failure. Is that OK?

---

Can the caller do anything if this fails? If not, I think we're good here. 

Maybe we'd like to print the error in the HTML as well?

---

I'm mostly concerned with functions that expect to continue business-as-usual. When things go sideways, we want to stop and hoist the "read the logs, please!" banner, not continue waiting for things to clear up by themselves (which they won't).

---

onIceCandidate is called twice, both by functions that are event handlers, and the only point of having the function is to get the reference to the pc that sent the event, which seems like the event should have info about (does it?)

Can we rewrite this function so that it can be directly assigned as the event handler?

It should call out to some global "something went wrong" routine that splashes a banner on top of the page and stops trying to do more things until user intervenes.

---

How about having a `<div id="errorMessages"></div>` where we append errors? We use a similar (but simpler) thing in other samples. Maybe expand on that? 

I agree about the problems with `onIceCandidate` but I choose not to change that now. Let me investigate if we can get the pc from the event and maybe eliminate this function as well.